### DISCUSSION TITLE
SQL Questions : Deleting the loan history after a certain period of time.

### DISCUSSION THREAD

**Discussion by meteos77** [textbrowser/biblioteq#286] [Q&A] [answered]
I would like to be able to delete the loan history after a certain period of time.
return_date + 4 months => history deleted

for example on May 1,  2023 : I want to delete the books returned before 31/01/2023.
                      on May 15, 2023 : I want to delete the books returned before 02/15/2023.

at each loading of BQ, can he check this data and do the necessary if needed?
Is there a way to manage this automatically?

For the moment I have created the following 2 queries :
SELECT * FROM member_history WHERE returned_date LIKE ('%/%/2022')

I think that this request works well for the deletion - can it create an adverse effect?
`DELETE FROM member_history WHERE returned_date LIKE ('%/%/2022')`

can you help me, thank you in advance.

**Comment by textbrowser** [comment]
Yet another reason SQLite is not a proper database engine.

https://www.sqlite.org/datatype3.html#date_and_time_datatype

So, what can you do? I will not answer this directly. Instead, examples.

Show items which are between specific dates.

SELECT * FROM member_history WHERE SUBSTR(reserved_date, 7, 4) || SUBSTR(reserved_date, 1, 2) || SUBSTR(reserved_date, 4, 2) BETWEEN '20230401' AND '20230425';

**Comment by textbrowser** [comment]
Show items which are between relative dates.

SELECT * FROM member_history WHERE SUBSTR(reserved_date, 7, 4) || '-' || SUBSTR(reserved_date, 1, 2) || '-' || SUBSTR(reserved_date, 4, 2) BETWEEN (SELECT DATE('now')) AND (SELECT DATE('now', 'start of month', '+25 day'));

**Comment by textbrowser** [comment]
You may tweak these to discover items which are within specific date intervals or relative date intervals.

**Comment by textbrowser** [comment]
Always run a SELECT query before committing to a DELETE.

**Comment by textbrowser** [comment]
Automatically? Kind of.

![bq-1](https://user-images.githubusercontent.com/10701156/233792053-7d391010-4f60-4c6f-bc0a-4c80974c9f78.png)

**Reply by meteos77** [reply]
I like to see complicated examples, it helps to learn   :  || '-' || ???

**Comment by meteos77** [comment]
thank you for all these explanations,

I'll deconstruct your examples to understand step by step.

**Comment by textbrowser** [comment]
Copy the database and test the queries first. Good luck.

**Reply by meteos77** [reply]
Thank you, I will keep you informed of my progress.

**Comment by meteos77** [comment]
45 days after the document is returned, the history is removed. //45 jours après le retour du document, l'historique est enlevé.

`DELETE FROM member_history WHERE myoid IN (SELECT myoid FROM member_history  WHERE SUBSTR(returned_date, 7, 4) || '-' || SUBSTR(returned_date, 1, 2) || '-' || SUBSTR(returned_date, 4, 2) BETWEEN (SELECT DATE ('2000-01-01'))  AND (SELECT DATE ('now' ,'-45 day')));`

**Comment by meteos77** [comment]
1 year after the last contribution the inactive member is deleted (calculation: current year -1 years, which is less than the result).
ex : 2023 - 1 = 20222 so 2021 is deleted.

1 an après la dernière cotisation le membre inactif est supprimé (calcul : année actuelle -1 ans , ce qui est inférieur au résultat).
ex : 2023 - 1 = 20222 donc 2021 est supprimé.

`DELETE FROM member WHERE memberid IN (SELECT memberid FROM member WHERE SUBSTR(expiration_date, 7, 4)  < (SELECT SUBSTR(DATE('now', '-1 year'), 1, 4)));`

**Comment by meteos77** [comment]
Thanks for the examples and tips.

Last little question
![Member's_Reservations_History-coche-Donotrecordreservationhistory](https://user-images.githubusercontent.com/56431039/233835208-7bdd660a-7f97-48c7-962e-b5af92597c31.png)


In the panel, there is a checkbox to not retain the history (Do not record reservation history) but I can't do anything on it ?

**Comment by textbrowser** [comment, marked as answer]
The option is available for real patrons. PostgreSQL.

**Reply by meteos77** [reply]
Ok PostgreSQL has more possibilities :-)

**Comment by textbrowser** [comment]
That widget is tagged with a dynamic tool tip. Added more text too.

**Comment by textbrowser** [comment]
Administrators and patrons don't exist in SQLite so the option doesn't apply.
